core/textile/backup.go

```diff
@@ -0,0 +1,266 @@
+		if item.IsDir {
+			p := strings.Join([]string{path, item.Name}, "/")
+
+			n, err := tc.backupBucketFiles(ctx, bucket, p)
+			if err != nil {
+				return 0, err
+			}
+
+			count += n
+			continue
+		}
+
+		// parallelize the backups
+		go func(path string) {
+			defer wg.Done()
+
+			if err = tc.backupFile(ctx, bucket, path); err != nil {
+				errc <- err
+				return
+			}
+
+			count += 1
+		}(item.Path)
+	}
+
+	wg.Wait()
+
+	return count, nil
+}
+
+// meta, key and share with me thread ids
+func (tc *textileClient) userThreadIds(ctx context.Context, bucket Bucket) ([]thread.ID, error) {
+	dbIds := make([]thread.ID, 0)
+
+	// key
+	bucketThreadID, err := bucket.GetThreadID(ctx)
+	if err != nil {
+		return nil, err
+	}
+	dbIds = append(dbIds, *bucketThreadID)
+
+	// shared with me
+	publicShareId, err := tc.getPublicShareThread(ctx)
+	if err != nil {
+		return nil, err
+	}
+	dbIds = append(dbIds, publicShareId)
+
+	return dbIds, nil
+}
+
+// replicate meta, key and shared with me threads
+func (tc *textileClient) replicateThreadsToHub(ctx context.Context, bucket Bucket) (count int, err error) {
+	replicatedDbIds := make([]thread.ID, 0)
+
+	// poor man's transactionality
+	defer func() {
+		if err != nil {
+			for _, dbId := range replicatedDbIds {
+				err = tc.DereplicateThreadFromHub(ctx, &dbId)
+				if err != nil {
+					log.Error(fmt.Sprintf("failed to dereplicate thread (dbId=%+v b.Slug=%+v)", dbId, bucket.Slug()), err)
+				}
+			}
+		}
+	}()
+
+	userThreadIds, err := tc.userThreadIds(ctx, bucket)
+	if err != nil {
+		return 0, err
+	}
+
+	for _, dbId := range userThreadIds {
+		if err = tc.ReplicateThreadToHub(ctx, &dbId); err != nil {
+			return 0, err
+		}
+
+		replicatedDbIds = append(replicatedDbIds, dbId)
+	}
+
+	return len(replicatedDbIds), nil
+}
+
+// backup the bucket
+func (tc *textileClient) BackupBucket(ctx context.Context, bucket Bucket) (count int, err error) {
+
+	count, err = tc.backupBucketFiles(ctx, bucket, "")
+	if err != nil {
+		return 0, nil
+	}
+
+	if _, err = tc.replicateThreadsToHub(ctx, bucket); err != nil {
+		return 0, nil
+	}
+
+	return count, nil
+}
+
+// unbackup a single file
+func (tc *textileClient) unbackupFile(ctx context.Context, bucket Bucket, path string) (err error) {
+
+	if err = tc.unsetMirrorFileBackup(ctx, path, bucket.Slug()); err != nil {
+		log.Error(fmt.Sprintf("error unsetting mirror file as backup (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	if err = tc.deleteFileFromHub(ctx, bucket, path); err != nil {
+		log.Error(fmt.Sprintf("error backuping up file with reader (path=%+v b.Slug=%+v)", path, bucket.Slug()), err)
+		return err
+	}
+
+	return nil
+}
+
+// unbackup all files in a bucket
+func (tc *textileClient) unbackupBucketFiles(ctx context.Context, bucket Bucket, path string) (int, error) {
+	var wg sync.WaitGroup
+	var count int
+
+	dir, err := bucket.ListDirectory(ctx, "")
+	if err != nil {
+		return 0, err
+	}
+
+	wg.Add(len(dir.Item.Items))
```

Comment:
Not sure if this is right. We are just adding one item to the working group (since this is not inside the for) ,so this is just going to continue past `wg.Wait()` after the first item does the `wg.Done()`.

Reply:
@dmerrill6 

Thank you for spotting this! This code was incorrect anyway because there is `.textile` and `.textileseed`. :-)

ref https://github.com/FleekHQ/space-daemon/pull/191/files/593541272f2e1edb98679e8020351e5a6ecb5232#diff-b580f3c0539be5a72155b66b76313a90R232